### PR TITLE
Add support for fine-tuning Chronos

### PR DESCRIPTION
**Pull request by abdulfatir**
*Description of changes:* Dummy PR for easy reviewing. Main PR: https://github.com/autogluon/autogluon/pull/4608

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: timeseries/src/autogluon/timeseries/models/chronos/model.py [hunk 1]
```diff
@@ -108,9 +110,28 @@ class ChronosModel(AbstractTimeSeriesModel):
...
+        If True, validation will be performed during fine-tuning to select the best checkpoint.
+        Setting this argument to True may result in slower fine-tuning.
+    shuffle_buffer_size : int, default = 10000
+        The size of the shuffle buffer to shuffle the data during fine-tuning. If None, shuffling will
+        be turned off.
+    eval_max_items : int, default = 100
+        The maximum number of randomly-sampled time series to use from the validation set for evaluation
+        during fine-tuning. If None, the entire validation dataset will be used.
+    trainer_kwargs : dict, optional
+        Extra keyword arguments passed to ``transformers.TrainingArguments``
```
Comment (shchur) [inline timeseries/src/autogluon/timeseries/models/chronos/model.py]: Would it make sense to fold all of these arguments into a `trainer_kwargs` dictionary? So instead of using the `fine_tune_*` prefix, the users could specify the fine-tuning config with something like following:
```python
"Chronos": {"fine_tune": True, "trainer_kwargs": {"lr": 1e-4, "batch_size": 16}},
```
This would provide a clear separation between the "standard" Chronos hyperparameters and those used only during fine-tuning.

File: timeseries/src/autogluon/timeseries/models/chronos/model.py [hunk 2]
```diff
@@ -108,9 +110,28 @@ class ChronosModel(AbstractTimeSeriesModel):
...
         Number of samples used during inference
+    fine_tune : bool, default = False
+        If True, the pretrained model will be fine-tuned
+    fine_tune_lr: float, default = 0.0001
+        The learning rate used for fine-tuning
+    fine_tune_steps : int, default = 5000
+        The number of gradient update steps to fine-tune for
+    fine_tune_batch_size : int, default = 16
+        The batch size to use for fine-tuning
+    eval_during_fine_tune : bool, default = False
```
Comment (shchur) [inline timeseries/src/autogluon/timeseries/models/chronos/model.py:124]: Nit (and maybe a bad idea): does it make sense to merge `eval_during_fine_tune` and `eval_max_items` into a single argument `num_val_items` that can take values 
- 0 - validation disabled (default)
- int - up to this many randomly selected items are used for validation
- None - all items are used for validation

File: timeseries/src/autogluon/timeseries/models/chronos/model.py [hunk 3]
```diff
@@ -273,15 +304,195 @@ def persist(self) -> "ChronosModel":
...
+        from .pipeline import ChronosBoltPipeline
+        from .pipeline.utils import ChronosFineTuningDataset, EvaluateAndSaveFinalStepCallback, TimeLimitCallback
+
+        # verbosity < 3: all logs and warnings from transformers will be suppressed
+        # verbosity >= 3: progress bar and loss logs will be logged
+        # verbosity 4: everything will be logged
+        verbosity = kwargs.get("verbosity", 2)
+        for logger_name in logging.root.manager.loggerDict:
+            if "transformers" in logger_name:
+                pl_logger = logging.getLogger(logger_name)
```
Comment (shchur) [inline timeseries/src/autogluon/timeseries/models/chronos/model.py]: Nit: `pl_logger` -> `transformers_logger`? `pl` used to stand for PyTorch Lightning before

File: timeseries/src/autogluon/timeseries/models/chronos/model.py [hunk 4]
```diff
@@ -273,15 +304,195 @@ def persist(self) -> "ChronosModel":
...
         self._check_fit_params()
-        self.time_limit = time_limit
+
+        fine_tune_args = self._get_model_params()
+        do_fine_tune = fine_tune_args["fine_tune"]
+        eval_during_fine_tune = val_data is not None and fine_tune_args["eval_during_fine_tune"]
+
+        if time_limit is not None:
+            fine_tune_time_limit = time_limit // 2 if do_fine_tune else 0
+            self.time_limit = time_limit - fine_tune_time_limit  # inference time budget
```
Comment (shchur) [inline timeseries/src/autogluon/timeseries/models/chronos/model.py]: Let's measure the actual fine-tuning time and use the remainder as the inference time limit?
```python
start_time = time.monotonic()
... # do fine-tuning
if time_limit is not None:
    self.time_limit = time_limit - (time.monotonic() - start_time)
```

File: timeseries/src/autogluon/timeseries/models/chronos/pipeline/utils.py [hunk 5]
```diff
@@ -1,14 +1,210 @@
...
+        if self.mode == "training":
+            iterable = self._create_training_data(self.gluonts_dataset)
+        elif self.mode == "validation":
+            iterable = self._create_validation_data(self.gluonts_dataset)
+
+        format_transform_fn = self.to_chronos_format if self.tokenizer is not None else self.to_chronos_bolt_format
+        for entry in iterable:
+            yield format_transform_fn(entry)
+
+    def shuffle(self, shuffle_buffer_size: int | None = None):
```
Comment (shchur) [inline timeseries/src/autogluon/timeseries/models/chronos/pipeline/utils.py]: IIRC `|` is not supported in Python 3.9. Can you please use `Union` or `from __future__ import annotations`?

File: timeseries/src/autogluon/timeseries/models/chronos/model.py (hunk 1 above)
Comment (abdulfatir) [reply timeseries/src/autogluon/timeseries/models/chronos/model.py]: The reason I went with this setup is:

- Only using `trainer_kwargs` will make things a little bit opaque. I expect this to only be used by advanced users. 
- In this way, we're exposing some core parameters such as steps, learning rate and batch size and advanced users can also provide additional `trainer_kwargs`, if needed. 
- The default parameter names in HF `TrainingArguments` are a bit elaborate and need knowledge of how `transformers` library works for one to be able to specify them. For example, training batch size is called `per_device_train_batch_size`.

File: timeseries/src/autogluon/timeseries/models/chronos/pipeline/utils.py (hunk 5 above)
Comment (canerturkmen) [reply timeseries/src/autogluon/timeseries/models/chronos/pipeline/utils.py]: in this case, rather `Optional[int]` ?

File: timeseries/src/autogluon/timeseries/models/chronos/model.py (hunk 2 above)
Comment (abdulfatir) [reply timeseries/src/autogluon/timeseries/models/chronos/model.py:124]: I like that things are separated. Unless there's a good reason to change, I would keep this as is.

File: timeseries/src/autogluon/timeseries/models/chronos/model.py (hunk 3 above)
Comment (abdulfatir) [reply timeseries/src/autogluon/timeseries/models/chronos/model.py]: Oops.

File: timeseries/src/autogluon/timeseries/models/chronos/model.py (hunk 4 above)
Comment (abdulfatir) [reply timeseries/src/autogluon/timeseries/models/chronos/model.py]: Done.

File: timeseries/src/autogluon/timeseries/models/chronos/pipeline/utils.py (hunk 5 above)
Comment (abdulfatir) [reply timeseries/src/autogluon/timeseries/models/chronos/pipeline/utils.py]: Done.
